ldelf/ta_elf_rel.c

```diff
@@ -78,63 +89,122 @@ static bool __resolve_sym(struct ta_elf *elf, unsigned int st_bind,
 	return true;
 }
 
-static TEE_Result resolve_sym_helper(uint32_t hash, const char *name,
-				     vaddr_t *val, struct ta_elf *elf,
-				     bool weak_ok)
+static bool check_found_sym(struct ta_elf *elf, const char *name,
+			    vaddr_t *val, bool weak_ok,
+			    size_t n, bool is_32)
 {
+	Elf32_Sym *sym32 = NULL;
+	Elf64_Sym *sym64 = NULL;
+
+	if (n >= elf->num_dynsyms)
+		err(TEE_ERROR_BAD_FORMAT, "Index out of range");
 	/*
-	 * Using uint32_t here for convenience because both Elf64_Word
-	 * and Elf32_Word are 32-bit types
+	 * We're loading values from sym[] which later
+	 * will be used to load something.
+	 * => Spectre V1 pattern, need to cap the index
+	 * against speculation.
 	 */
-	uint32_t *hashtab = elf->hashtab;
-	uint32_t nbuckets = hashtab[0];
-	uint32_t nchains = hashtab[1];
-	uint32_t *bucket = &hashtab[2];
-	uint32_t *chain = &bucket[nbuckets];
-	size_t n = 0;
+	n = confine_array_index(n, elf->num_dynsyms);
+
+	if (is_32) {
+		sym32 = elf->dynsymtab;
+		if (sym_compare(elf,
+				ELF32_ST_BIND(sym32[n].st_info),
+				ELF32_ST_TYPE(sym32[n].st_info),
+				sym32[n].st_shndx,
+				sym32[n].st_name,
+				sym32[n].st_value, name, val, weak_ok))
+			return true;
+	} else {
+		sym64 = elf->dynsymtab;
+		if (sym_compare(elf,
+				ELF64_ST_BIND(sym64[n].st_info),
+				ELF64_ST_TYPE(sym64[n].st_info),
+				sym64[n].st_shndx,
+				sym64[n].st_name,
+				sym64[n].st_value, name, val, weak_ok))
+			return true;
+	}
 
-	if (elf->is_32bit) {
-		Elf32_Sym *sym = elf->dynsymtab;
+	return false;
+}
 
-		for (n = bucket[hash % nbuckets]; n; n = chain[n]) {
-			if (n >= nchains || n >= elf->num_dynsyms)
-				err(TEE_ERROR_BAD_FORMAT,
-				    "Index out of range");
-			/*
-			 * We're loading values from sym[] which later
-			 * will be used to load something.
-			 * => Spectre V1 pattern, need to cap the index
-			 * against speculation.
-			 */
-			n = confine_array_index(n, elf->num_dynsyms);
-			if (__resolve_sym(elf,
-					  ELF32_ST_BIND(sym[n].st_info),
-					  ELF32_ST_TYPE(sym[n].st_info),
-					  sym[n].st_shndx,
-					  sym[n].st_name,
-					  sym[n].st_value, name, val, weak_ok))
-				return TEE_SUCCESS;
+static TEE_Result resolve_sym_helper(const char *name, vaddr_t *val,
+				     struct ta_elf *elf, bool weak_ok)
+{
+	uint32_t n = 0;
+	uint32_t hash = 0;
+
+	if (elf->gnu_hashtab) {
+		/*
+		 * Using uint32_t here for convenience because both Elf64_Word
+		 * and Elf32_Word are 32-bit types
+		 */
+		uint32_t *hashtab = elf->gnu_hashtab;
+		uint32_t nbuckets = hashtab[0];
+		uint32_t symskip = hashtab[1];
+		uint32_t bloom_size = hashtab[2];
+		uint32_t *bucket = NULL;
+		uint32_t *chain = NULL;
+		uint32_t hashval = 0;
+
+		hash = gnu_hash(name);
+
+		if (elf->is_32bit) {
+			uint32_t *bloom = (void *)&hashtab[4];
```

Comment:
Cast not needed, here and at line 161. 
Do you prefer to keep them for consistency with the 64bit case instruction sequence?

Reply:
No, useless is useless ;) I'll fix.